// react-cosmos-shared2/dist/FixtureLoader/useValue/useCreateFixtureState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCreateFixtureState = useCreateFixtureState;

var _lodash = require("lodash");

var _react = _interopRequireDefault(require("react"));

var _fixtureState = require("../../fixtureState");

var _FixtureContext = require("../FixtureContext");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function useCreateFixtureState(inputName, defaultValue) {
  var _React$useContext = _react.default.useContext(_FixtureContext.FixtureContext),
      setFixtureState = _React$useContext.setFixtureState;

  _react.default.useEffect(function () {
    // The fixture state for this value is (re)created in two situations:
    // 1. Initially: No corresponding fixture state value is found
    // 2: Default value change: Current value is reset to new default value
    setFixtureState(function (prevFsState) {
      var fsValuePair = (0, _fixtureState.findFixtureStateValue)(prevFsState, inputName);

      if (fsValuePair && fsValueExtendsBaseValue(fsValuePair.defaultValue, defaultValue)) {
        return prevFsState;
      }

      return _objectSpread({}, prevFsState, {
        values: _objectSpread({}, prevFsState.values, _defineProperty({}, inputName, {
          defaultValue: (0, _fixtureState.createValue)(defaultValue),
          currentValue: (0, _fixtureState.createValue)(defaultValue)
        }))
      });
    });
  }, [setFixtureState, inputName, defaultValue]);
}

function fsValueExtendsBaseValue(fsValue, baseValue) {
  return (0, _lodash.isEqual)(baseValue, (0, _fixtureState.extendWithValue)(baseValue, fsValue));
}
