// react-cosmos/dist/dom/ErrorCatch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorCatch = void 0;

var _react = _interopRequireDefault(require("react"));

var _lodash = require("lodash");

var _react2 = require("react-cosmos-shared2/react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ErrorCatch =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ErrorCatch, _React$Component);

  function ErrorCatch() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, ErrorCatch);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(ErrorCatch)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      error: null
    });

    return _this;
  }

  _createClass(ErrorCatch, [{
    key: "componentDidCatch",
    value: function componentDidCatch(error, info) {
      this.setState({
        error: "".concat(error.message, "\n").concat(info.componentStack)
      });
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      // A change in fixture (children) or fixture state signifies that the
      // problem that caused the current error might've been solved. If the error
      // persists, it will organically trigger the error state again in the next
      // update
      if (this.state.error && (fixtureChanged(this.props.children, prevProps.children) || fixtureStateChanged(this.props.fixtureState, prevProps.fixtureState))) {
        this.setState({
          error: null
        });
        this.props.onErrorReset();
      }
    }
  }, {
    key: "render",
    value: function render() {
      return this.state.error ? this.renderError(this.state.error) : this.props.children;
    }
  }, {
    key: "renderError",
    value: function renderError(error) {
      // Don't render error details here in dev mode because react-error-overlay
      // takes care of it in a nicer way. We used to render both for a while but
      // it proved annoying. react-error-overlay has a slight delay and seeing
      // the same error reported twice feels clumsy.
      if (__DEV__) {
        return null;
      } // In static exports, however, where react-error-overlay is missing,
      // rendering plain error details is superior to showing a blank screen.


      return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("p", null, _react.default.createElement("strong", null, "Ouch, something wrong!")), _react.default.createElement("pre", null, error), _react.default.createElement("p", null, "Check console for more info."));
    }
  }]);

  return ErrorCatch;
}(_react.default.Component);

exports.ErrorCatch = ErrorCatch;

function fixtureChanged(f1, f2) {
  return !(0, _react2.areNodesEqual)(f1, f2, true);
}

function fixtureStateChanged(fS1, fS2) {
  return !(0, _lodash.isEqual)(fS1, fS2);
}
