// react-cosmos/dist/plugins/webpack/webpackConfig/devServer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDevWebpackConfig = getDevWebpackConfig;

var _path = _interopRequireDefault(require("path"));

var _webpack = require("../cosmosConfig/webpack");

var _htmlPlugin = require("./htmlPlugin");

var _shared = require("./shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function getDevWebpackConfig(_x, _x2) {
  return _getDevWebpackConfig.apply(this, arguments);
}

function _getDevWebpackConfig() {
  _getDevWebpackConfig = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(cosmosConfig, userWebpack) {
    var baseWebpackConfig;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _shared.getUserWebpackConfig)(cosmosConfig, userWebpack);

          case 2:
            baseWebpackConfig = _context.sent;
            return _context.abrupt("return", _objectSpread({}, baseWebpackConfig, {
              entry: getEntry(cosmosConfig),
              output: getOutput(cosmosConfig),
              module: _objectSpread({}, baseWebpackConfig.module, {
                rules: getRules(baseWebpackConfig)
              }),
              resolve: (0, _shared.resolveLocalReactDeps)(cosmosConfig, baseWebpackConfig),
              plugins: getPlugins(cosmosConfig, baseWebpackConfig, userWebpack)
            }));

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _getDevWebpackConfig.apply(this, arguments);
}

function getEntry(cosmosConfig) {
  var _createWebpackCosmosC = (0, _webpack.createWebpackCosmosConfig)(cosmosConfig),
      hotReload = _createWebpackCosmosC.hotReload; // The React devtools hook needs to be imported before any other module that
  // might import React


  var devtoolsHook = (0, _shared.resolveClientPath)('reactDevtoolsHook');
  var clientIndex = (0, _shared.resolveClientPath)('index');
  return hotReload ? [devtoolsHook, getHotMiddlewareEntry(), clientIndex] : [devtoolsHook, clientIndex];
}

function getOutput(_ref) {
  var publicUrl = _ref.publicUrl;
  var filename = '[name].js';
  return {
    // Setting path to `/` in development (where files are saved in memory and
    // not on disk) is a weird required for old webpack versions
    path: '/',
    filename: filename,
    publicPath: publicUrl,
    // Enable click-to-open source in react-error-overlay
    devtoolModuleFilenameTemplate: function devtoolModuleFilenameTemplate(info) {
      return _path.default.resolve(info.absoluteResourcePath).replace(/\\/g, '/');
    }
  };
}

function getRules(baseWebpackConfig) {
  var existingRules = baseWebpackConfig.module && baseWebpackConfig.module.rules || [];
  return [].concat(_toConsumableArray(existingRules), [(0, _shared.getUserDepsLoaderRule)()]);
}

function getPlugins(cosmosConfig, baseWebpackConfig, userWebpack) {
  var existingPlugins = baseWebpackConfig.plugins || [];
  var globalsPlugin = (0, _shared.getGlobalsPlugin)(cosmosConfig, userWebpack, true);
  var noEmitErrorsPlugin = new userWebpack.NoEmitOnErrorsPlugin();
  var plugins = [].concat(_toConsumableArray(existingPlugins), [globalsPlugin, noEmitErrorsPlugin]);

  var _createWebpackCosmosC2 = (0, _webpack.createWebpackCosmosConfig)(cosmosConfig),
      hotReload = _createWebpackCosmosC2.hotReload;

  if (hotReload && !(0, _shared.hasPlugin)(plugins, 'HotModuleReplacementPlugin')) {
    var hmrPlugin = new userWebpack.HotModuleReplacementPlugin();
    plugins = [].concat(_toConsumableArray(plugins), [hmrPlugin]);
  }

  return (0, _htmlPlugin.ensureHtmlWebackPlugin)(cosmosConfig, plugins);
}

function getHotMiddlewareEntry() {
  var clientPath = require.resolve('@skidding/webpack-hot-middleware/client');

  return "".concat(clientPath, "?reload=true&overlay=false");
}
