// react-cosmos/dist/plugins/webpack/webpackConfig/export.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getExportWebpackConfig = getExportWebpackConfig;

var _path = _interopRequireDefault(require("path"));

var _shared = require("../../../shared/shared");

var _htmlPlugin = require("./htmlPlugin");

var _shared2 = require("./shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function getExportWebpackConfig(_x, _x2) {
  return _getExportWebpackConfig.apply(this, arguments);
}

function _getExportWebpackConfig() {
  _getExportWebpackConfig = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(cosmosConfig, userWebpack) {
    var baseWebpackConfig;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _shared2.getUserWebpackConfig)(cosmosConfig, userWebpack);

          case 2:
            baseWebpackConfig = _context.sent;
            return _context.abrupt("return", _objectSpread({}, baseWebpackConfig, {
              entry: getEntry(),
              output: getOutput(cosmosConfig),
              module: _objectSpread({}, baseWebpackConfig.module, {
                rules: getRules(baseWebpackConfig)
              }),
              resolve: (0, _shared2.resolveLocalReactDeps)(cosmosConfig, baseWebpackConfig),
              plugins: getPlugins(cosmosConfig, baseWebpackConfig, userWebpack)
            }));

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _getExportWebpackConfig.apply(this, arguments);
}

function getEntry() {
  // The React devtools hook needs to be imported before any other module that
  // might import React
  var devtoolsHook = (0, _shared2.resolveClientPath)('reactDevtoolsHook');
  var clientIndex = (0, _shared2.resolveClientPath)('index');
  return [devtoolsHook, clientIndex];
}

function getOutput(_ref) {
  var exportPath = _ref.exportPath,
      publicUrl = _ref.publicUrl;
  var filename = '[name].js';
  return {
    path: _path.default.resolve(exportPath, (0, _shared.removeLeadingSlash)(publicUrl)),
    filename: filename,
    publicPath: publicUrl
  };
}

function getRules(baseWebpackConfig) {
  var existingRules = baseWebpackConfig.module && baseWebpackConfig.module.rules || [];
  return [].concat(_toConsumableArray(existingRules), [(0, _shared2.getUserDepsLoaderRule)()]);
}

function getPlugins(cosmosConfig, baseWebpackConfig, userWebpack) {
  var existingPlugins = baseWebpackConfig.plugins || [];
  var globalsPlugin = (0, _shared2.getGlobalsPlugin)(cosmosConfig, userWebpack, false);
  var noEmitErrorsPlugin = new userWebpack.NoEmitOnErrorsPlugin();
  return (0, _htmlPlugin.ensureHtmlWebackPlugin)(cosmosConfig, [].concat(_toConsumableArray(existingPlugins), [globalsPlugin, noEmitErrorsPlugin]));
}
